conbench/entities/history.py

```diff
@@ -570,3 +583,82 @@ def _calculate_z_score(
         z_score = z_score * -1
 
     return z_score
+
+
+class _AutomatedDetector(ABC):
+    @abstractmethod
+    def detect(self, df: pd.DataFrame) -> pd.Series:
+        """Detect outliers and distribution shifts in historical data
+
+        For a particular detection algorithm, should take a dataframe of the same
+        sort as `_add_rolling_stats_columns_to_df`, and return that dataframe with
+        two additional columns:
+
+        - `is_step` (bool): Is this point the start of a new segment?
+        - `is_outlier` (bool): Is this point an outlier that should be ignored?
+        """
+
+
+class _ZScoreAutomatedDetector(_AutomatedDetector):
+    def detect(self, df: pd.DataFrame, z_score_threshold=5.0) -> pd.DataFrame:
+        tmp_df = copy.deepcopy(df)
+
+        # skip computation if no history
+        if df.shape[0] == 0:
+            tmp_df["is_step"] = pd.Series([], dtype=bool)
+            tmp_df["is_outlier"] = pd.Series([], dtype=bool)
+            return tmp_df
+
+        # pandas likes the data to be sorted
+        tmp_df.sort_values(
+            [
+                "case_id",
+                "context_id",
+                "hash",
+                "repository",
+                "timestamp",
+                "result_timestamp",
+            ],
+            inplace=True,
+            ignore_index=True,
+        )
+
+        # split / apply
+        out_group_df_list = []
+        for _, group_df in tmp_df.groupby(
+            ["case_id", "context_id", "hash", "repository"]
+        ):
+            # clean copy will only get result columns
+            out_group_df = copy.deepcopy(group_df)
+
+            group_df["mean_diff"] = group_df["mean"].diff()
+            mean_diff_clipped = group_df.mean_diff[
+                (group_df.mean_diff > group_df.mean_diff.quantile(0.05))
+                & (group_df.mean_diff < group_df.mean_diff.quantile(0.95))
+            ]
+            group_df["z_score"] = (
+                group_df.mean_diff - mean_diff_clipped.mean()
+            ) / mean_diff_clipped.std()
```

Comment:
Just so we're all clear: this analysis uses an expanding window, correct? So when we look for outliers in mean-diff, we consider the distribution of mean-diffs since the beginning of time.

Notably, this is a different strategy from the lookback z-score method. We had a conversation on why that method needed a max window size [here](https://github.com/conbench/conbench/pull/636#issuecomment-1412422072). Basically, if we didn't have one, we'd risk the distribution having too high a variance (without a LOT of careful pruning).

Here I think I agree with you that that matters less. I don't think the mean-diff distribution necessarily would change a lot over time. In fact, we're kind of relying on a similar property in the lookback z-score method, where we assume the spread of the means doesn't change much over time. (Plus you're clipping outliers so that should help a little more.)

Reply:
It's not even expanding here, it's just the straight mean of everything minus the clipped points. I think I could do it as expanding or rolling if I inserted nans instead of filtering. That would be more reproducible than this, which depends entirely on what data gets sent. 

Practically speaking, though, none of these are likely to vary much at all, so I'll see if I can make this more reproducible without overcomplicating too much.

Reply:
Switched these to rolling using the same `Config.DISTRIBUTION_COMMITS`. In practice I don't think what that value is will matter much here, but rolling generally will make results more reproducible when called with different slices of data. And it doesn't peek at the future.

Reply:
Nice!